02-intermediate-ruby/lazy-with-let.md

```diff
@@ -0,0 +1,180 @@
+# Being Lazy With Let
+## Learning Goals
+You should be able to: 
+- DRY up your code with let & before.
+- Explain how `let` is "lazy" in Minitest.
+
+In your Minitest code you will often find yourself creating instances over and over again in each test case like this:
+
+```ruby
+it "must have a name" do
+  player = Scrabble::Player.new("Ada Lovelace")
+  
+  # expectations go here
+end
+```
+
+In most test cases we're creating a new identical instance of `Player` to work with.  This doesn't seem DRY.  
+
+## Solution 1:  Before Blocks
+
+One solution is to use Minitest's `before` block. 
+
+```ruby
+describe Player do
+  before do
+    @player = Scrabble::Player.new("Ada Lovelace")
+  end
+
+  it "must respond to name" do
+    @dplauer.must_respond_to :name
+  end
+  ...
+end
+```
+
+So in the above example the `before` block runs before each test and provides an instance variable `@player` which we can use in our tests.  The problem with before blocks is that they **always** run before each test case.  Even if we don't want to use `@player` in a few test cases, like for instance if we want to try creating a Player with no arguments the before block will still run.  
+
+## Let
+
+Let is another way to DRY up your code.  With Let we create a block which can be referenced in our test cases.
+
+
+```ruby
+# player_spec.rb from Scrabble Project
+
+describe Player do
+  let (:player)  { 						 # <-- The name used
+  	Scrabble::Player.new("Ada Lovelace") # <-- The value returned by 'player'
+  }
+
+
+	it "has a name" do
+		player.name.must_equal "Watson"
+    end
+end
+```
+
+The `let` statement creates a new method with the given name (in this case `player`) that executes the code in the curly braces and caches the result for later use.  That's why it's often called "lazy" the code inside the block doesn't execute until the let is first used, in this case when we do `player.name.must_equal "Watson"`.  So the let runs for each test case that uses it and doesn't run when it isn't needed.  Neat!
+
+### Proof that let is lazy
+
+In this example using `before` the statement "Creating a player named Ada Lovelace" prints twice times showing that before always runs.  
```

Comment:
I might clarify that there are two tests (aka `it` blocks) and that is why we see it twice